src/workerd/api/tcp.h

```diff
@@ -0,0 +1,130 @@
+  void abortRead() override;
+  void getsockopt(int level, int option, void* value, uint* length) override;
+  void setsockopt(int level, int option, const void* value, uint length) override;
+
+  void getsockname(struct sockaddr* addr, uint* length) override;
+  void getpeername(struct sockaddr* addr, uint* length) override;
+
+  // AsyncInputStream methods:
+  kj::Promise<size_t> read(void* buffer, size_t minBytes, size_t maxBytes) override;
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
+
+  // AsyncOutputStream methods:
+  kj::Promise<void> write(const void* buffer, size_t size) override;
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override;
+
+  kj::Promise<void> whenWriteDisconnected() override;
+
+private:
+  // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
+  // stored stream (if the `inner` promise completed).
+  template <typename T>
+  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
+    KJ_IF_MAYBE(e, error) {
+      kj::throwRecoverableException(kj::mv(*e));
+    }
+
+    KJ_IF_MAYBE(s, completedInner) {
+      return f(s);
+    } else {
+      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
+        completedInner = kj::mv(stream);
+        return f(&KJ_ASSERT_NONNULL(completedInner));
+      });
+    }
+  }
+
+  // Stored io stream once the `inner` promise completes.
+  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
+  // Stored error if any of the operations throw.
+  kj::Maybe<kj::Exception> error;
+  // A promise holding a stream that will be available at some point in the future. Typically once
+  // a connection is established.
+  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
+};
+
+struct SocketOptions: public jsg::Object {
+  // TODO(later): TCP socket options will live here.
+  SocketOptions() {};
+public:
+  JSG_RESOURCE_TYPE(SocketOptions, CompatibilityFlags::Reader flags) { }
+};
+
+class Socket: public jsg::Object {
+public:
+  Socket(kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise);
+
+  kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise;
+
+  jsg::Ref<ReadableStream> getReadable() { return KJ_ASSERT_NONNULL(readable).addRef(); }
+  jsg::Ref<WritableStream> getWritable() { return KJ_ASSERT_NONNULL(writable).addRef(); }
+  jsg::Promise<void> getClosed() {
+    // TODO: Right now this promise won't complete when the remote end closes the socket.
+    //       Do we need to wrap ReadableStream/WritableStream to make this work or is there a
+    //       better way?
+    auto& context = IoContext::current();
+    return context.awaitIo(kj::mv(closeFulfiller->promise));
+  }
+
+  jsg::Promise<void> close(jsg::Lock& js);
+  // Closes the socket connection.
+
+  JSG_RESOURCE_TYPE(Socket, CompatibilityFlags::Reader flags) {
+    if (flags.getJsgPropertyOnPrototypeTemplate()) {
```

Comment:
For a brand new type checking this flag is unnecessary. This object does not have a `constructor()` implementation. If we are not expecting to allow folks to subclass this thing (e.g. `new MySocket extends Socket {...}`, then we can expose the getters/setters using just `JSG_READONLY_INSTANCE_PROPERTY`.